### PR TITLE
Restore support for GCC-4.8.

### LINKED ISSUE #798: Reconsider the decision to drop gcc-4.8
The TensorFlow folks still support gcc-4.8, and it would be nice to not break their builds all the time.

This is hard though, because nohmann/json refuses to compile with gcc-4.8, so we may have to disable the storage/ subdirectory with older compilers.

### PR DESCRIPTION
**Pull request by coryan**
With the latest version of nlohmann::json we can support GCC-4.8 again.

This fixes #798.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1731)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,14 +27,13 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
 
-    # gcc-4.8 C++11 support has bugs preventing us from using it.
-    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.9)
+    # gcc-4.8 is the first release that claims C++11 support.
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.8)
         message(
             FATAL_ERROR
                 "GCC version must be at least 4.9. Older versions"
```
Comment (devbww) [inline CMakeLists.txt]: s/4.9/4.8/

File: google/cloud/firestore/field_path.cc [hunk 2]
```diff
@@ -63,10 +61,33 @@ FieldPath FieldPath::Append(FieldPath const& field_path) const {
 }
 
 std::string FieldPath::ToApiRepr() const {
+#if defined(__GNUC__) && __GNUC__ == 4 && __GNUC_MINOR__ == 8
```
Comment (devbww) [inline google/cloud/firestore/field_path.cc]: Do we like the "defined(\_\_GNUC\_\_)"?  If it wasn't defined, "\_\_GNUC\_\_ == 4" would be false anyway.  Or would that make this expression gratuitously different from other uses where that doesn't apply?

File: google/cloud/firestore/field_path.cc [hunk 3]
```diff
@@ -63,10 +61,33 @@ FieldPath FieldPath::Append(FieldPath const& field_path) const {
...
+    }
+    for (auto const& c : part) {
+      if (c != '_' and not std::isalnum(c)) {
+        return false;
+      }
+    }
+    return true;
+  };
+#else
+  auto is_simple = [](std::string const& part) {
```
Comment (devbww) [inline google/cloud/firestore/field_path.cc]: "is_simple_field_name" perhaps?

File: google/cloud/storage/status_or.h [hunk 4]
```diff
@@ -181,8 +185,10 @@ class StatusOr final {
   Status& status() & { return status_; }
   Status const& status() const& { return status_; }
   Status&& status() && { return std::move(status_); }
+#if GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
   Status const&& status() const&& { return std::move(status_); }
-  //@}
+#endif  // GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+        //@}
```
Comment (devbww) [inline google/cloud/storage/status_or.h]: Is the indentation change significant?  (And below.)

File: google/cloud/firestore/field_path.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/firestore/field_path.cc]: There are a few other uses of `defined(__GNUC__)` in this file, so I would prefer to keep as-is for now. Later we can change all of them, I have no opinion on which style is "better".

Having said that, if we find ourselves testing for compiler names and versions in a lot of places we are doing something wrong.

File: CMakeLists.txt (hunk 1 above)
Comment (coryan) [reply CMakeLists.txt]: Fixed.

File: google/cloud/firestore/field_path.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/firestore/field_path.cc]: Fixed.  On second thought I removed the code using regex, seems like overkill anyway, happy to restore it if you think it is important.

File: google/cloud/storage/status_or.h (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/status_or.h]: No, clang-format just did that, we could insert a blank line or reorder the functions to make it look better.  Frankly I just let clang-format do its thing, most of the time it is good, and I rather worry about other things. I will change it if you think it is important, of course.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/firestore/field_path.cc (hunk 3 above)
Comment (devbww) [reply google/cloud/firestore/field_path.cc]: No ... I prefer the non-conditional too.

File: google/cloud/storage/status_or.h (hunk 4 above)
Comment (devbww) [reply google/cloud/storage/status_or.h]: I'm happy to let clang-format have its way (mostly) too.

**Comment by devbww** [comment]
It looks like there were some check errors, but I can't seem to see them.

**Comment by coryan** [comment]
I fixed the permissions, you should be able to see the Kokoro builds now. I also fixed the build. For extra
fun it looks like we had a VM crash or BSOD in the middle of our build yay!

**Pull request merged**